### PR TITLE
[stats]: Subgroup cluster stats() into lb-stats, endpoint-stats, config-update-stats and "the rest"(will be renamed to upstream-stats)

### PR DESCRIPTION
**Pull request by stevenzzzz**
Commit Message: Subgroup cluster stats() into lb-stats, endpoint-stats, config-update-stats and "the rest"(will be renamed to upstream-stats)
Additional Description:
See more description and live example in https://github.com/envoyproxy/envoy/issues/23575.

With lots of clusters and route-tables in a cloud proxy, we are seeing tons of RAM been spent on stats while most of the stats are never inc-ed due to traffic pattern(or long tail). We are thinking that we can lazy init cluster stats() so that the RAM is only allocated when it's required.

To achieve that we need to have finer grained stats group, e.g. configUpdateStats() are frequently updated by config management server, while upstream_xxx are only required when there is traffic for the cluster, for this sub-group we can save RAM by lazy init it.

This is the subgroups I plan to have for clusters.

/**

All cluster config update related stats.
*/
#define ALL_CLUSTER_CONFIG_UPDATE_STATS(COUNTER, GAUGE, HISTOGRAM,
TEXT_READOUT, STATNAME)
COUNTER(assignment_stale)
COUNTER(assignment_timeout_received)
COUNTER(update_attempt)
COUNTER(update_empty)
COUNTER(update_failure)
COUNTER(update_no_rebuild)
COUNTER(update_success)
GAUGE(version, NeverImport)
/**

All cluster endpoints related stats.
*/
#define ALL_CLUSTER_ENDPOINT_STATS(COUNTER, GAUGE, HISTOGRAM, TEXT_READOUT,
STATNAME)
GAUGE(max_host_weight, NeverImport)
COUNTER(membership_change)
GAUGE(membership_degraded, NeverImport)
GAUGE(membership_excluded, NeverImport)
GAUGE(membership_healthy, NeverImport)
GAUGE(membership_total, NeverImport)
/**

All cluster loadbalancing related stats.
*/
#define ALL_CLUSTER_LB_STATS(COUNTER, GAUGE, HISTOGRAM, TEXT_READOUT,
STATNAME)
COUNTER(lb_healthy_panic)
COUNTER(lb_local_cluster_not_ok)
COUNTER(lb_recalculate_zone_structures)
COUNTER(lb_subsets_created)
COUNTER(lb_subsets_fallback)
COUNTER(lb_subsets_fallback_panic)
COUNTER(lb_subsets_removed)
COUNTER(lb_subsets_selected)
COUNTER(lb_zone_cluster_too_small)
COUNTER(lb_zone_no_capacity_left)
COUNTER(lb_zone_number_differs)
COUNTER(lb_zone_routing_all_directly)
COUNTER(lb_zone_routing_cross_zone)
COUNTER(lb_zone_routing_sampled)
GAUGE(lb_subsets_active, Accumulate)
/**

All cluster stats. https://github.com/see stats_macros.h
*/
#define ALL_CLUSTER_STATS(COUNTER, GAUGE, HISTOGRAM, TEXT_READOUT, STATNAME)
COUNTER(bind_errors)
COUNTER(original_dst_host_invalid)
COUNTER(retry_or_shadow_abandoned)
COUNTER(upstream_cx_close_notify)
COUNTER(upstream_cx_connect_attempts_exceeded)
COUNTER(upstream_cx_connect_fail)
COUNTER(upstream_cx_connect_timeout)
COUNTER(upstream_cx_connect_with_0_rtt)
COUNTER(upstream_cx_destroy)
COUNTER(upstream_cx_destroy_local)
COUNTER(upstream_cx_destroy_local_with_active_rq)
COUNTER(upstream_cx_destroy_remote)
COUNTER(upstream_cx_destroy_remote_with_active_rq)
COUNTER(upstream_cx_destroy_with_active_rq)
COUNTER(upstream_cx_http1_total)
COUNTER(upstream_cx_http2_total)
COUNTER(upstream_cx_http3_total)
COUNTER(upstream_cx_idle_timeout)
COUNTER(upstream_cx_max_duration_reached)
COUNTER(upstream_cx_max_requests)
COUNTER(upstream_cx_none_healthy)
COUNTER(upstream_cx_overflow)
COUNTER(upstream_cx_pool_overflow)
COUNTER(upstream_cx_protocol_error)
COUNTER(upstream_cx_rx_bytes_total)
COUNTER(upstream_cx_total)
COUNTER(upstream_cx_tx_bytes_total)
COUNTER(upstream_flow_control_backed_up_total)
COUNTER(upstream_flow_control_drained_total)
COUNTER(upstream_flow_control_paused_reading_total)
COUNTER(upstream_flow_control_resumed_reading_total)
COUNTER(upstream_internal_redirect_failed_total)
COUNTER(upstream_internal_redirect_succeeded_total)
COUNTER(upstream_rq_cancelled)
COUNTER(upstream_rq_completed)
COUNTER(upstream_rq_maintenance_mode)
COUNTER(upstream_rq_max_duration_reached)
COUNTER(upstream_rq_pending_failure_eject)
COUNTER(upstream_rq_pending_overflow)
COUNTER(upstream_rq_pending_total)
COUNTER(upstream_rq_0rtt)
COUNTER(upstream_rq_per_try_timeout)
COUNTER(upstream_rq_per_try_idle_timeout)
COUNTER(upstream_rq_retry)
COUNTER(upstream_rq_retry_backoff_exponential)
COUNTER(upstream_rq_retry_backoff_ratelimited)
COUNTER(upstream_rq_retry_limit_exceeded)
COUNTER(upstream_rq_retry_overflow)
COUNTER(upstream_rq_retry_success)
COUNTER(upstream_rq_rx_reset)
COUNTER(upstream_rq_timeout)
COUNTER(upstream_rq_total)
COUNTER(upstream_rq_tx_reset)
COUNTER(upstream_http3_broken)
GAUGE(upstream_cx_active, Accumulate)
GAUGE(upstream_cx_rx_bytes_buffered, Accumulate)
GAUGE(upstream_cx_tx_bytes_buffered, Accumulate)
GAUGE(upstream_rq_active, Accumulate)
GAUGE(upstream_rq_pending_active, Accumulate)
HISTOGRAM(upstream_cx_connect_ms, Milliseconds)
HISTOGRAM(upstream_cx_length_ms, Milliseconds)

Risk Level: LOW, should be a no-behavior-change CL.

Testing: N/A existing stats test should prove that there is no behavior change.
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: envoy/upstream/upstream.h
Comment (jmarantz) [inline envoy/upstream/upstream.h:570]: This comment should describe the current state of affairs assuming this PR is checked in, not express a plan.

AFAICT stats_ no longer exists for upstream; it's now fully partitioned so we can just document that.

File: envoy/upstream/upstream.h
Comment (jmarantz) [inline envoy/upstream/upstream.h:620]: I think maybe we should rename this macro and related call-sites because this does not represent all the stats anymore.

File: envoy/upstream/upstream.h
Comment (jmarantz) [inline envoy/upstream/upstream.h:1035]: I realize this phrase 'strongly named' pre-existed, but I'm not sure what it means.

It's possible I came up with phrase and i still don't know what it means :)

Do you? If not, can we just remove this here & below?

For ClusterLbStats maybe use 'load-balancer-related stats for this cluster'.

File: source/common/upstream/cluster_manager_impl.h
Comment (jmarantz) [inline source/common/upstream/cluster_manager_impl.h:328]: do we still use ClusterStatNames anywhere?

**Review by jmarantz — COMMENTED** [review]
looks good basically; just a few nits.

Can you merge main?

File: envoy/upstream/upstream.h
Comment (stevenzzzz) [reply envoy/upstream/upstream.h:620]: I did that in another CL. See https://github.com/envoyproxy/envoy/pull/23921/files

File: envoy/upstream/upstream.h
Comment (stevenzzzz) [reply envoy/upstream/upstream.h]: I decided to rename that field in the next PR. 
I think the comment here is kinda accurate, updated the "upstream" one to reflect the "current" status.

File: envoy/upstream/upstream.h
Comment (jmarantz) [reply envoy/upstream/upstream.h:620]: can we focus on just making this one coherent? and getting it in?

File: envoy/upstream/upstream.h
Comment (stevenzzzz) [reply envoy/upstream/upstream.h]: Done.

File: source/common/upstream/cluster_manager_impl.h
Comment (stevenzzzz) [reply source/common/upstream/cluster_manager_impl.h:329]: I am not sure. 
I saw it only used to instantiate the ClusterStats in upstream_impl.cc.

**Review by stevenzzzz — COMMENTED** [review]
Thanks! PTAL.

**Comment by jmarantz** [comment]
confused about how this stacks on the other PR -- this works on its own and appears coherent, right? Also that one is a draft and this one looks close to being ready.

**Comment by stevenzzzz** [comment]
@jmarantz let's commit this one and ignore the other one in the chain.  (https://github.com/envoyproxy/envoy/pull/23523)

File: source/common/upstream/cluster_manager_impl.h
Comment (jmarantz) [reply source/common/upstream/cluster_manager_impl.h:329]: can you comment it out and see what breaks?

I'd like this PR to just be a repartitioning without adding any overlap.

File: test/common/upstream/load_balancer_fuzz_base.h
Comment (jmarantz) [inline test/common/upstream/load_balancer_fuzz_base.h:44]: hm. I guess you didn't need to change the variable name from stat_names_ to lb_stat_names_ here. You could have just left it at stat_names_ (still changing the type. Ditto for lb_stats_. That would make this PR a lot smaller I think.

Comment (stevenzzzz) [reply test/common/upstream/load_balancer_fuzz_base.h:44]: IIUC, there is a statsname for every StatStruct. So we have now 4 {Lb, Endpoint, ConfigUpdate, }StatNames definitions.

File: test/common/upstream/load_balancer_impl_test.cc
Comment (jmarantz) [inline test/common/upstream/load_balancer_impl_test.cc:1962]: is this change related?

BTW it's fine to set stats in a test. It's only in prod that I have a concern. In any case this change probably should go in a different PR, no?

File: test/common/upstream/zone_aware_load_balancer_fuzz_base.cc
Comment (jmarantz) [inline test/common/upstream/zone_aware_load_balancer_fuzz_base.cc:51]: here too.

Actually if the reason you woke to this is that (a) these stats were no longer accessible from this point and so coudn't be written and (b) that had no impact on the tests, then maybe send just the deletion of these max_host_weight_set() calls into a separate PR? That should be easy to review (deleting a bunch of superfluous test lines).

One concern I might have is if setting these values wakes up some code we need to be testable.

**Review by jmarantz — COMMENTED** [review]
lgtm modulo minor rename request, plus the superflous changes to set host weight.

thank you!

**Review by stevenzzzz — COMMENTED** [review]
thanks!

**Comment by jmarantz** [comment]
great. this is OK with me but you might want to throw the removal of stats_.max_host_weight_.set into a separate PR which should be easy to review & merge first. That will avoid having a senior maintainer ask the same question :)

this lgtm to me other than that, and your format fail :)

**Review by jmarantz — APPROVED** [review]
Thanks Xin.

Stephen please let me know if you need more context or reach out to the PR author direclty.

File: envoy/upstream/upstream.h
Comment (stevenzzzz) [reply envoy/upstream/upstream.h]: done.

File: test/common/upstream/zone_aware_load_balancer_fuzz_base.cc
Comment (stevenzzzz) [reply test/common/upstream/zone_aware_load_balancer_fuzz_base.cc:51]: removing the max_host_weight_ references in tests with PR: https://github.com/envoyproxy/envoy/pull/23936

**Comment by jmarantz** [comment]
I think we should wait for a non-google maintainer to review. If Stephan can't do it we can get a non-Google non-senior maintainer and then Yan can give final approval.

**Comment by jmarantz** [comment]
Also PTAL @ clang-tidy failure

File: source/common/upstream/upstream_impl.cc
Comment (yanavlasov) [inline source/common/upstream/upstream_impl.cc:837]:
```suggestion
  return {stat_names, scope};
```

File: test/common/upstream/load_balancer_simulation_test.cc
Comment (yanavlasov) [inline test/common/upstream/load_balancer_simulation_test.cc:107]: I know this is not your change, but we need to mollify preexisting clang-tidy

```suggestion
class DISABLED_SimulationTest : public testing::Test { // NOLINT(readability-identifier-naming)
```

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: test/common/upstream/load_balancer_simulation_test.cc
Comment (stevenzzzz) [reply test/common/upstream/load_balancer_simulation_test.cc]: Thanks!

**Review by stevenzzzz — COMMENTED** [review]
thanks!

File: envoy/upstream/upstream.h
Comment (ggreenway) [inline envoy/upstream/upstream.h:597]: I think these should be further sub-divided by functionality. lb_subset, zone_aware, and then all the rest. Both of those features are unused in many configs so they could all be omitted.

**Review by ggreenway — CHANGES_REQUESTED** [review]

File: envoy/upstream/upstream.h
Comment (stevenzzzz) [reply envoy/upstream/upstream.h:597]: IIUC, the zone_aware stats are the "base" for subset_lb loadbalancers, due to the following: 
```
class EdfLoadBalancerBase : public ZoneAwareLoadBalancerBase
```
so the lb-related stats (subset, and zone-aware) will be created/required at the same time.

Comment (ggreenway) [reply envoy/upstream/upstream.h:597]: But I think the zone-awareness isn't used in many cases, and those stats will remain zero'd (at least I think that's how it works). So if they are separate and lazy-created, I think they won't get created in many configs.

Comment (stevenzzzz) [reply envoy/upstream/upstream.h:597]: our of the 5 lb-impls used in subset_lb, only RingHashLoadBalancer and MagLevLoadBalancer are NOT inherited from zoneawareLoadBalancer. But I see your point there tho. It requires some refactoring work: we'd better instantiate those stats objs from within each lb-object, which triggers even more delta.

do you mind if I do it incrementally?

Comment (ggreenway) [reply envoy/upstream/upstream.h:597]: Incrementally sounds good. Do you want to land this as-is and do it in a subsequent PR?

Comment (stevenzzzz) [reply envoy/upstream/upstream.h:597]: yeah. it'd be much cleaner.

**Review by ggreenway — APPROVED** [review]

**Comment by stevenzzzz** [comment]
Thanks for working on my PR!

**Pull request merged**
